### DISCUSSION TITLE
Local method overloading

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/csharplang#4909] [Q&A] [answered]
It seems like local methods don't support overloading. Is this a known limitation?

It might not matter in most cases, but gets really confusing for top level programs. And considering top level programs are intended to ease learning curve of C#, this definitely doesn't serve that goal. 

![image](https://user-images.githubusercontent.com/6609929/124983804-1e40ee80-e041-11eb-9b83-be7ff3f50e05.png)

CC: @KathleenDollard ([as you asked](https://youtu.be/J0LTqwTojlA?t=1142), my 2 cents on making top level program a default console template)

**Comment by CyrusNajmabadi** [comment]
This is a discussion for csharplang as it involves potentially changing the language.  I would generally be in support of this, though there may be hidden complexity that makes this difficult.

**Comment by jnm2** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/discussions/3112

**Comment by HaloFour** [comment]
I thought forbidding overloading of local functions was a very explicit language design decision (and that if several members of the language design team could reverse time they might forbid overloading in general in the C# language)?

**Reply by maxkoshevoi** [reply]
Could you please elaborate on "overloading being a bad thing for C# language"? Haven't heard that one before.

More on the topic of the discussion: Since method overloading is part of C# and top level program [might be default console template in .Net 6](https://youtu.be/J0LTqwTojlA?t=1142), it would make sense to be able to overload local functions as well (or make compiler place top level functions outside `Main` method, so that they can be overloaded)

**Reply by 333fred** [reply]
My general view is "Pick either implicit conversions or method overloading. If you pick both, you're going to have a hell of a time with overload resolution."

**Reply by CyrusNajmabadi** [reply]
My rule on thumb is: language changes are easy, unless they involve overload resolution. In which case, 80%+ of the time is spent just dealing with that.

**Reply by CyrusNajmabadi** [reply]
> Didn't heard that one before.

It's not that it's bad per se. It just occupies a disproportionate amount of time versus the value it provides imo

**Reply by DavidArno** [reply]
> My general view is "Pick either implicit conversions or method overloading. If you pick both, you're going to have a hell of a time with overload resolution."

So when we add discriminated unions to the mix, potentially with implicit conversion between the union's case types and those actual types, this will will allow the team to descend to the next level of hell I'm guessing. 😬

**Reply by 333fred** [reply]
> > My general view is "Pick either implicit conversions or method overloading. If you pick both, you're going to have a hell of a time with overload resolution."
> 
> So when we add discriminated unions to the mix, potentially with implicit conversion between the union's case types and those actual types, this will will allow the team to descend to the next level of hell I'm guessing. 😬 

If we do add that as something other than a user-defined conversion, you're absolutely right. I doubt we will though, you'd just destructure the union. Unless you're thinking of some scenario I'm not imagining right now?
